src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -0,0 +1,346 @@
+
+    def validate(self, node, options=None):
+        '''Perform various checks to ensure that it is valid to apply the
+        NemoArrayRange2LoopTrans transformation to the supplied PSyIR Node.
+
+        :param node: the node that is being checked.
+        :type node: :py:class:`psyclone.psyir.nodes.Range`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        :raises TransformationError: if the node argument is not a \
+            Range, if the Range node is not part of an ArrayReference, \
+            if the Range node is not the outermost Range node of the \
+            ArrayReference or if that ArrayReference does not \
+            consitute the left hand side of an Assignment node.
+
+        '''
+        # Am I Range node?
+        if not isinstance(node, Range):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be a PSyIR Range, but "
+                "found '{0}'.".format(type(node).__name__))
+        # Am I within an array reference?
+        if not node.parent or not isinstance(node.parent, ArrayReference):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node, but found '{0}'.".format(type(node.parent).__name__))
+        array_ref = node.parent
+        # Is the array reference within an assignment?
+        if not array_ref.parent or not isinstance(array_ref.parent,
+                                                  Assignment):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within an Assignment node, but found '{0}'."
+                .format(type(array_ref.parent).__name__))
+        assignment = array_ref.parent
+        # Is the array reference the lhs of the assignment?
+        if assignment.lhs is not array_ref:
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within the left-hand-side of an Assignment "
+                "node, but it is on the right-hand-side.")
+        # Does the rhs of the assignment have any operations that
+        # return arrays as this is not currently supported?
+        for operation in assignment.rhs.walk(Operation):
+            # At the moment the only array valued operator is matmul
+            if operation.operator == BinaryOperation.Operator.MATMUL:
```

Comment:
Really we should have a list of array-valued operators but it's a small point and can be done later.

Reply:
Yes, that would be better. It is also possible that #685 and #688 would provide an alternative solution as we could determine what the return type is instead of keeping a list.